### PR TITLE
plugins

### PR DESCRIPTION
**Pull request by nicolasparada**
New plugins abstraction that enables a more Golang idiomatic way of writing plugins.

Docs have not been updated. Let's tag this and build a couple of actual plugins to test edge cases, and then let's update docs.

For now let's reference the plugins in the `plugin/testdata` directory.

CI is not yet configured on this repo 🤔 
Let's leave that out of the scope of this PR.
Tests do pass. And lint using `golangci-lint` too.

### CODE REVIEW

**Comment by nicolasparada** [comment]
For the time being, I'm getting a lot of:

```text
caught signal (SIGSEGV)
```

From fluent-bit but I don't know how to debug that kind of issues.

## For a linux/amd64 machine

Try running
```bash
make local-plugins
make fluentbit-run
```

You will see the error most of the time at start or when stopping fluentbit.
(You can repeat the second command if you didn't got the error as it don't happen always).

## For a non linux/amd64 machine

Run:
```bash
make docker-image
make docker-image-run
```

Same history.
(You can repeat the 2nd cmd).

File: plugin/testdata/plugins.conf [hunk 1]
```diff
@@ -0,0 +1,5 @@
+[PLUGINS]
+    Path /fluent-bit/etc/go-test-input-plugin.so
```
Comment (nicolasparada) [inline plugin/testdata/plugins.conf:2]: I still think the plugin name and description should be placed here and not in the plugin code itself.
If you share a plugin by just downloading an `.so` file there is no way of knowing the plugin name 🤔

File: plugin/cshared.go [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+				fmt.Fprintf(os.Stderr, "run: %s\n", err)
+				return output.FLB_ERROR
+			}
+
+			return output.FLB_OK
+		default:
+		}
+
+		var entry []any
+		err := dec.Decode(&entry)
```
Comment (nicolasparada) [inline plugin/cshared.go:185]: I don't know if all these assertions of `[]any{fTime, map[string]string}` could potentially broke something.

I saw the first input plugin implementation also checked for `int64` besides `fTime`.
But I guess that was because the msgpack writer was exposed directly and the plugin author could have been writen anything really. ANd it was counting for the case the plugin was sending it as unix seconds.

Comment (niedbalski) [reply plugin/cshared.go:185]: No, internally it won't have a different format. I think your assumption is correct. []any{fTime, map[string]string}

File: plugin/testdata/plugins.conf (hunk 1 above)
Comment (niedbalski) [reply plugin/testdata/plugins.conf:2]: +1 we should extend this optionally for allowing a name to be passed. lets keep it as is for now.

File: plugin/time.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+	"time"
+
+	"github.com/ugorji/go/codec"
+)
+
+var _ codec.BytesExt = (*fTime)(nil)
+
+// fTime implements codec.BytesExt to handle custom (de)serialization of types to/from []byte.
+// It is used by codecs (e.g. binc, msgpack, simple) which do custom serialization of the types.
+type fTime time.Time
```
Comment (niedbalski) [inline plugin/time.go:15]: fTime is confusing as a type name.

File: plugin/time.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	return b
+}
+
+func (*fTime) ReadExt(dst interface{}, src []byte) {
+	ft, ok := dst.(*fTime)
+	if !ok {
+		panic(fmt.Sprintf("unexpected fluent time type %T", dst))
+	}
+
+	if len(src) != 8 {
```
Comment (niedbalski) [inline plugin/time.go:40]: make this a const

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,26 @@
+FROM golang:latest AS builder
```
Comment (niedbalski) [inline Dockerfile:1]: let's don't version this in the repository and make it a example (could be under docs/examples or so).

Comment (nicolasparada) [reply Dockerfile]: Moved it to the tests under testdata.

File: plugin/time.go (hunk 4 above)
Comment (nicolasparada) [reply plugin/time.go]: Done.

File: plugin/time.go (hunk 3 above)
Comment (nicolasparada) [reply plugin/time.go:17]: Any proposal?
`flbTime`?

Comment (cosmo0920) [reply plugin/time.go:17]: `flbTime` or `fbTime` is better.

File: plugin/testdata/plugins.conf (hunk 1 above)
Comment (cosmo0920) [reply plugin/testdata/plugins.conf:2]: -1 I don’t think we should provide plugin name and description extensibility in config file. IMO, I don’t want to introduce this complexity but this is only my two cents.

Comment (pwhelan) [reply plugin/testdata/plugins.conf:2]: -1. I agree with Hiroshi, there is no reason we should force people to guess what the internal name of a plug-in is when trying to load it.

The only use I can think of is to override a plugins name when loading it. That might be useful but a bit out of scope for the moment.

Comment (nicolasparada) [reply plugin/testdata/plugins.conf:2]: I see it the opposite. You have to guess it right now.

I propose that the plugin itself has no name or description at all in its code.

When you load it you have to name it.

File: input/encoder.go [hunk 6]
```diff
@@ -59,7 +58,7 @@ func NewEncoder() *FLBEncoder {
 	enc := new(FLBEncoder)
 	enc.handle = new(codec.MsgpackHandle)
 	enc.handle.WriteExt = true
-	enc.handle.SetBytesExt(reflect.TypeOf(FLBTime{}), 0, &FLBTime{})
+	_ = enc.handle.SetBytesExt(reflect.TypeOf(FLBTime{}), 0, &FLBTime{})
```
Comment (niedbalski) [inline input/encoder.go:61]: don't shadow errors

File: plugin/cshared.go [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"reflect"
+	"time"
+	"unsafe"
+
+	"github.com/fluent/fluent-bit-go/input"
```
Comment (niedbalski) [inline plugin/cshared.go:18]: aren't we referring to calyptia?

File: plugin/cshared.go [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+			fmt.Fprintf(os.Stderr, "encode: %s\n", err)
+			return input.FLB_ERROR
+		}
+
+		cdata := C.CBytes(b)
+
+		*data = cdata
+		*csize = C.size_t(len(b))
+
+		// C.free(unsafe.Pointer(cdata))
```
Comment (niedbalski) [inline plugin/cshared.go:120]: I think this is been freed by the plugin mechanism, but worth doing a plugin run under valgrind (?)

File: plugin/cshared.go [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+	}
+
+	return output.FLB_OK
+}
+
+//export FLBPluginExit
+func FLBPluginExit() int {
+	log.Printf("calling FLBPluginExit(): name=%q\n", theName)
+
+	if runCancel != nil {
```
Comment (niedbalski) [inline plugin/cshared.go:255]: will this ever be nil?

File: plugin/plugin_test.go [hunk 10]
```diff
@@ -0,0 +1,132 @@
+package plugin
```
Comment (niedbalski) [inline plugin/plugin_test.go:1]: we'll likely superseed this with bats testing but good for now.

File: plugin/plugin_test.go [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+	wantNoErr(t, err)
+
+	lines := strings.Split(string(contents), "\n")
+
+	// after 5 seconds of fluentbit running, there should be at least one record.
+	if d := len(lines); d < 1 {
+		t.Fatalf("expected at least 1 lines, got %d", d)
+	}
+
+	// Input plugin sends:
```
Comment (niedbalski) [inline plugin/plugin_test.go:99]: remove if unused

File: plugin/testdata/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+RUN go mod download
+RUN go mod verify
+
+COPY . .
+
+RUN go build -trimpath -buildmode c-shared -o ./bin/go-test-input-plugin.so ./plugin/testdata/input/input.go
+RUN go build -trimpath -buildmode c-shared -o ./bin/go-test-output-plugin.so ./plugin/testdata/output/output.go
+
+FROM ghcr.io/calyptia/enterprise/advanced:main
```
Comment (niedbalski) [inline plugin/testdata/Dockerfile:16]: I think this got renamed to calyptia/calyptia-fluent-bit/advanced:main

File: plugin/testdata/plugins.conf (hunk 1 above)
Comment (niedbalski) [reply plugin/testdata/plugins.conf:2]: > -1. I agree with Hiroshi, there is no reason we should force people to guess what the internal name of a plug-in is when trying to load it.
> 
> The only use I can think of is to override a plugins name when loading it. That might be useful but a bit out of scope for the moment.

I agree with this approach. Lets keep it as defined by the plugin for the time being.

File: output/decoder.go [hunk 13]
```diff
@@ -60,7 +60,7 @@ func NewDecoder(data unsafe.Pointer, length int) *FLBDecoder {
 
 	dec := new(FLBDecoder)
 	dec.handle = new(codec.MsgpackHandle)
-	dec.handle.SetBytesExt(reflect.TypeOf(FLBTime{}), 0, &FLBTime{})
+	_ = dec.handle.SetBytesExt(reflect.TypeOf(FLBTime{}), 0, &FLBTime{})
```
Comment (niedbalski) [inline output/decoder.go:63]: don't shadow errors

File: plugin/testdata/Dockerfile (hunk 12 above)
Comment (nicolasparada) [reply plugin/testdata/Dockerfile:16]: The repo was renamed but I see that the images have the old name still 🤔

File: plugin/plugin_test.go (hunk 11 above)
Comment (nicolasparada) [reply plugin/plugin_test.go:99]: I would like to keep it.

File: plugin/cshared.go (hunk 9 above)
Comment (nicolasparada) [reply plugin/cshared.go:255]: I don't know how the whole plugin lifecycle works from C side. I thought the plugin could be registered and exit for some reason right after. Or even initialized without any collection/flush and exited right after.
That's why I put it here to be safe in the case if was not initialized.

File: plugin/cshared.go (hunk 8 above)
Comment (nicolasparada) [reply plugin/cshared.go:120]: I need help to check those. That's something I don't understand at all.

File: input/encoder.go (hunk 6 above)
Comment (nicolasparada) [reply input/encoder.go:61]: There was no error handling, and I don't want to break change this package yet. This was just to pass linting. I'll put a code comment I think.

File: plugin/testdata/Dockerfile (hunk 12 above)
Comment (nicolasparada) [reply plugin/testdata/Dockerfile:16]: Nevermind. It did pull :)

File: plugin/cshared.go (hunk 7 above)
Comment (nicolasparada) [reply plugin/cshared.go:18]: That's a good question. This package is still named `github.com/fluent/fluent-bit-go`. See https://github.com/calyptia/fluent-bit-go/blob/b3af4fd037f50d243b0b5e61ca5834cf3b36e70c/go.mod#L1

Shall I rename it? 🤔 

Will we continue on this fork as a separate thing, or will we merge it to the fluent one at some point?

File: plugin/testdata/Dockerfile (hunk 12 above)
Comment (nicolasparada) [reply plugin/testdata/Dockerfile]: Done

File: output/decoder.go (hunk 13 above)
Comment (nicolasparada) [reply output/decoder.go:64]: Added code comment as I don't want to break change this yet.
This was just to pass linting.
